File: android/ValleTPV/ValleTPVlib/src/main/java/com/valleapp/valletpvlib/CashlogyManager/ChangeAction.java
package com.valleapp.valletpvlib.CashlogyManager;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

public class ChangeAction extends CashlogyAction {

    private double importeAdmitido = 0.0;

    boolean isCancel = false;
    boolean isAceptar = true;
    boolean isSending = false;

    int denominacionMinima = 0;

    boolean esBloqueado = false;
     Map<Integer, Integer> denominacionesDisponibles = new HashMap<>();

    public ChangeAction(CashlogySocketManager socketManager) {
        super(socketManager);
    }

    @Override
    public void execute() {
        importeAdmitido = 0;
        denominacionMinima = 0;
        isAceptar = true;
        isCancel = false;
        esBloqueado = false;
        isSending = false;
        socketManager.sendCommand("#Y#");
    }

    @Override
    public void handleResponse(String comando, String response) {
        if (comando.startsWith("#Y")) {
            manejarRespuestaDenominaciones(response);
        } else if (comando.startsWith("#B")) {
            sendQCommand();
        } else if (comando.startsWith("#Q#")) {
            manejarRespuestaQ(response);
        } else if (comando.startsWith("#J#")) {
            manejarRespuestaJ(response);
        }else if (comando.startsWith("#U")){
            manejarDispensacionCambio();
        } else if (comando.startsWith("#P")) {
            manejarRespuestaP();
        }
    }



    private void sendQCommand() {
        new Handler(Looper.getMainLooper()).postDelayed(() ->
                socketManager.sendCommand("#Q#"), 200);  // Esperar 200ms antes de volver a enviar #Q#
    }

    private void sendJCommand() {
        new Handler(Looper.getMainLooper()).postDelayed(() ->
                socketManager.sendCommand("#J#"), 200);  // Esperar 300ms antes de enviar #J#
    }

    private void sendPCommand(Double importe) {
        int centimosADevolver = (int) Math.round(importe * 100);
        String comandoDispensar = "#P#" + centimosADevolver + "#0#0#0#";
        new Handler(Looper.getMainLooper()).postDelayed(() ->
                socketManager.sendCommand(comandoDispensar), 200);  // Enviar comando #P# después de un pequeño retraso
    }

    private void manejarRespuestaDenominaciones(String response) {
        try {
            String[] parts = response.split("#");
            if (parts.length < 3) {
                Log.e("CASHLOGY", "Respuesta de denominaciones no válida.");
                socketManager.notifyUI("CASHLOGY_ERR", "Respuesta de denominaciones no válida.");
                return;
            }

            // Parte b de la respuesta, que contiene los recicladores (antes del primer punto y coma)
            String recicladoresParte = parts[2].split(";")[0]; // Primer parte son los recicladores (monedas y billetes)
            String billetesParte = parts[2].split(";")[1]; // Segunda parte para billetes

            // Limpiar el mapa antes de rellenarlo
            denominacionesDisponibles.clear();

            // Separar las denominaciones y cantidades para monedas (primera parte)
            String[] recicladores = recicladoresParte.split(",");
            for (String reciclador : recicladores) {
                String[] denomination = reciclador.split(":");
                if (denomination.length == 2) {
                    int valorEnCentimos = Integer.parseInt(denomination[0]);
                    int cantidad = Integer.parseInt(denomination[1]);

                    // Convertir de céntimos a euros
                    denominacionesDisponibles.put(valorEnCentimos, cantidad);
                }
            }

            // Separar las denominaciones y cantidades para billetes (segunda parte)
            String[] billetes = billetesParte.split(",");
            for (String billete : billetes) {
                String[] denomination = billete.split(":");
                if (denomination.length == 2) {
                    int valorEnCentimos = Integer.parseInt(denomination[0]);
                    int cantidad = Integer.parseInt(denomination[1]);
                    denominacionesDisponibles.put(valorEnCentimos, cantidad);
                }
            }

            socketManager.sendCommand("#B#0#0#0#");

        } catch (Exception e) {
            Log.e("CASHLOGY", "Error al manejar la respuesta de denominaciones: " + response + " Error: " + e.getMessage(), e);
            socketManager.notifyUI("CASHLOGY_ERR", "Error al manejar la respuesta de denominaciones.");
        }
    }

    private void manejarRespuestaQ(String response) {
        manejarRespuestaImporteAdmitido(response);
        // Si la admisión aún no ha finalizado, continuar consultando el importe
        if (isAceptar) {
            sendQCommand();
        }  else {
            sendJCommand();
        }
    }

    private void manejarRespuestaJ(String response) {
        manejarRespuestaImporteAdmitido(response);
        calcularYDispensarCambio();
    }


    private void manejarRespuestaImporteAdmitido(String response) {
        String[] parts = response.split("#");
        if (parts.length >= 3) {
            double nuevoImporteAdmitido = Double.parseDouble(parts[2]) / 100;

            // Solo notificar a la UI si el importe recibido es diferente al guardado
            if (nuevoImporteAdmitido != importeAdmitido) {
                importeAdmitido =Double.parseDouble( String.format(Locale.getDefault(),"%.2f", nuevoImporteAdmitido));
                socketManager.notifyUI("CASHLOGY_IMPORTE_ADMITIDO", String.valueOf(importeAdmitido));


                // Convertir la nueva cantidad a céntimos
                int importeRestante = (int) Math.round(importeAdmitido * 100);

                // Filtrar denominaciones menores o iguales a la cantidad introducida y que no tengan valor 0
                Map<Integer, Integer> denominacionesFiltradas = new HashMap<>();
                for (Map.Entry<Integer, Integer> entry : denominacionesDisponibles.entrySet()) {
                    int denominacion = entry.getKey();
                    int cantidad = entry.getValue();

                    if (denominacion <= importeRestante && cantidad > 0) {
                        denominacionesFiltradas.put(denominacion, cantidad);
                    }
                }

                // Enviar las denominaciones disponibles a la UI
                notificarDenominacionesDisponiblesUI(denominacionesFiltradas);
            }

        }
    }

    private void notificarDenominacionesDisponiblesUI(Map<Integer, Integer> denominacionesDisponibles) {
        StringBuilder cantidadesDisponibles = new StringBuilder();

        for (Map.Entry<Integer, Integer> entry : denominacionesDisponibles.entrySet()) {
            int denominacion = entry.getKey();
            int cantidad = entry.getValue();
            cantidadesDisponibles.append(denominacion).append(":").append(cantidad).append(",");
        }

        // Eliminar la última coma si existe
        if (cantidadesDisponibles.length() > 0) {
            cantidadesDisponibles.setLength(cantidadesDisponibles.length() - 1);
        }

        // Enviar la información de las denominaciones a la UI
        socketManager.notifyUI("CASHLOGY_DENOMINACIONES_DISPONIBLES", cantidadesDisponibles.toString());
    }



    private void calcularYDispensarCambio() {
        if (isSending) {
            return;
        }

        isSending = true;
        if (isCancel) {
            if (importeAdmitido > 0) {
                sendPCommand(importeAdmitido);
            }else{
                socketManager.notifyUI("CASHLOGY_CAMBIO", "Operación cancelada.");
            }
        } else {
            sendUCommand();
        }

    }

    private void manejarRespuestaP() {
        if(!isCancel){
            socketManager.notifyUI("CASHLOGY_CAMBIO", "Operación cancelada.");
        }else {
            socketManager.notifyUI("CASHLOGY_CAMBIO", "Operación finalizada.");
        }
    }

    private void manejarDispensacionCambio() {
        double importeRestante = Double.parseDouble(String.format(Locale.getDefault(),
                "%.2f", importeAdmitido - (denominacionMinima / 100.0)));
        sendPCommand(importeRestante);
    }

    private void sendUCommand() {
        StringBuilder comandoBuilder = new StringBuilder("#U#");

        if (denominacionMinima < 500) {
            // Si la denominación mínima es menor a 500 céntimos (monedas)
            comandoBuilder.append(denominacionMinima).append(":1;");
        } else {
            // Si la denominación mínima es 500 céntimos o mayor (billetes)
            comandoBuilder.append(";").append(denominacionMinima).append(":1");
        }

        // Completar el comando con los parámetros adicionales
        comandoBuilder.append("#0#0#0#");

        // Enviar el comando construido
        socketManager.sendCommand(comandoBuilder.toString());
    }




    public void cancelar(){
        if(!esBloqueado) {
            esBloqueado = true;
            isCancel = true;
            isAceptar = false;
        }
    }

    public void cambiar(int denominacionMinima){
        if (importeAdmitido > 0) {
            if(!esBloqueado) {
                esBloqueado = true;
                isCancel = false;
                isAceptar = false;
                this.denominacionMinima = denominacionMinima;
            }
        }
    }
}

File: android/ValleTPV/ValleTPVlib/src/main/java/com/valleapp/valletpvlib/cashlogymanager/ChangeAction.kt
package com.valleapp.valletpvlib.cashlogymanager

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.valleapp.valletpvlib.tools.round

class ChangeAction(
    socketManager: CashlogySocketManager,
    uiHandler: Handler
) : CashlogyAction(socketManager, uiHandler) {

    private var importeAdmitido = 0.0
    private var isCancel = false
    private var isAceptar = true
    private var isSending = false
    private var denominacionMinima = 0
    private var esBloqueado = false
    private val denominacionesDisponibles = mutableMapOf<Int, Int>()

    private val handler: Handler = Handler(Looper.getMainLooper()) { msg ->
        val key = msg.data.getString("key")
        if (key == null  || key != "CASHLOGY_RESPONSE") return@Handler true

        val comando = msg.data.getString("instruccion")
        val response = msg.data.getString("value")

        if (comando != null && response != null ) {
            when {
                comando.startsWith("#Y") -> manejarRespuestaDenominaciones(response)
                comando.startsWith("#B") -> sendQCommand()
                comando.startsWith("#Q#") -> manejarRespuestaQ(response)
                comando.startsWith("#J#") -> manejarRespuestaJ(response)
                comando.startsWith("#U") -> manejarDispensacionCambio()
                comando.startsWith("#P") -> manejarRespuestaP()
            }
        }
        true
    }

    override fun execute() {
        importeAdmitido = 0.0
        denominacionMinima = 0
        isAceptar = true
        isCancel = false
        esBloqueado = false
        isSending = false
        socketManager.sendCommand("#Y#", handler)
    }

    private fun sendQCommand() {
        Handler(Looper.getMainLooper()).postDelayed({
                socketManager.sendCommand("#Q#", handler)
        }, 200)
    }

    private fun sendJCommand() {
        Handler(Looper.getMainLooper()).postDelayed({
                socketManager.sendCommand("#J#", handler)
        }, 200)
    }

    private fun sendPCommand(importe: Double) {
        val centimosADevolver = (importe * 100).toInt()
        val comandoDispensar = "#P#$centimosADevolver#0#0#0#"
        Handler(Looper.getMainLooper()).postDelayed({
                socketManager.sendCommand(comandoDispensar, handler)
        }, 200)
    }

    private fun manejarRespuestaDenominaciones(response: String) {
        try {
            Log.d("CASHLOGY", "Respuesta de denominaciones: $response")
            val parts = response.split("#")
            if (parts.size < 3) {
                Log.e("CASHLOGY", "Respuesta de denominaciones no válida.")
                notifyUI("CASHLOGY_ERR", "Respuesta de denominaciones no válida.")
                return
            }

            val recicladoresParte = parts[2].split(";")[0]
            val billetesParte = parts[2].split(";")[1]
            denominacionesDisponibles.clear()

            recicladoresParte.split(",").forEach { reciclador ->
                    val (valorEnCentimos, cantidad) = reciclador.split(":").map { it.toInt() }
                denominacionesDisponibles[valorEnCentimos] = cantidad
            }

            billetesParte.split(",").forEach { billete ->
                    val (valorEnCentimos, cantidad) = billete.split(":").map { it.toInt() }
                denominacionesDisponibles[valorEnCentimos] = cantidad
            }

            socketManager.sendCommand("#B#0#0#0#", handler)


        } catch (e: Exception) {
            Log.e("CASHLOGY", "Error al manejar la respuesta de denominaciones: $response Error: ${e.message}", e)
            notifyUI("CASHLOGY_ERR", "Error al manejar la respuesta de denominaciones.")
        }
    }

    private fun manejarRespuestaQ(response: String) {
        manejarRespuestaImporteAdmitido(response)
        if (isAceptar) {
            sendQCommand()
        } else {
            sendJCommand()
        }
    }

    private fun manejarRespuestaJ(response: String) {
        manejarRespuestaImporteAdmitido(response)
        calcularYDispensarCambio()
    }

    private fun manejarRespuestaImporteAdmitido(response: String) {
        val parts = response.split("#")
        if (parts.size >= 3) {
            val nuevoImporteAdmitido = (parts[2].toInt() / 100.0).round(2)

            if (nuevoImporteAdmitido != importeAdmitido) {
                importeAdmitido = nuevoImporteAdmitido
                notifyUI("CASHLOGY_IMPORTE_ADMITIDO", importeAdmitido.toString())

                val importeRestante = (importeAdmitido * 100).toInt()
                val denominacionesFiltradas = denominacionesDisponibles.filter { it.key <= importeRestante && it.value > 0 }
                notificarDenominacionesDisponiblesUI(denominacionesFiltradas)
            }
        }
    }

    private fun notificarDenominacionesDisponiblesUI(denominacionesDisponibles: Map<Int, Int>) {
        val cantidadesDisponibles = denominacionesDisponibles.entries.joinToString(",") { "${it.key}:${it.value}" }
        notifyUI("CASHLOGY_DENOMINACIONES_DISPONIBLES", cantidadesDisponibles)
    }

    private fun calcularYDispensarCambio() {
        if (isSending) return
        isSending = true

        if (isCancel) {
            if (importeAdmitido > 0) {
                sendPCommand(importeAdmitido)
            } else {
                notifyUI("CASHLOGY_CAMBIO", "Operación cancelada.")
            }
        } else {
            sendUCommand()
        }

        if (isCancel) {
            notifyUI("CASHLOGY_CAMBIO", "Operación cancelada.")
        } else {
            notifyUI("CASHLOGY_CAMBIO", "Operación finalizada.")
        }
    }

    private fun manejarRespuestaP() {
        notifyUI("CASHLOGY_CAMBIO", "Maquina lista para usaar")
    }

    private fun manejarDispensacionCambio() {
        val importeRestante = (importeAdmitido - denominacionMinima / 100.0).round(2)
        sendPCommand(importeRestante)
        notifyUI("CASHLOGY_CAMBIO", "Operación finalizada.")
    }

    private fun sendUCommand() {
        val comandoBuilder = StringBuilder("#U#")
        if (denominacionMinima < 500) {
            comandoBuilder.append("$denominacionMinima:1;")
        } else {
            comandoBuilder.append(";").append("$denominacionMinima:1")
        }
        comandoBuilder.append("#0#0#0#")
        socketManager.sendCommand(comandoBuilder.toString(), handler)
    }

    fun cancelar() {
        if (!esBloqueado) {
            esBloqueado = true
            isCancel = true
            isAceptar = false
        }
    }

    fun cambiar(denominacionMinima: Int) {
        if (importeAdmitido > 0 && !esBloqueado) {
            esBloqueado = true
            isCancel = false
            isAceptar = false
            this.denominacionMinima = denominacionMinima
        }
    }
}
